### PR TITLE
Theme Toggling Feature 

### PR DESCRIPTION
**Pull request by ayanaar**
This PR addresses the last step in the process of implementing theme support for Pontoon. It introduces the capability for users to seamlessly switch between the platform's existing dark theme and the newly added light theme (Issue #2936 ). 

### Changes:
- **New UI Element**: Added buttons in the settings allowing users to select their preferred theme: Dark, Light, or System Default.
- **Theme Persistence**: User theme preference is stored in their profile, ensuring a consistent experience across sessions.
- **Dynamic Theme Application**: Leveraging JavaScript, the theme is applied dynamically without requiring a page reload.
- **Fallback Mechanism**: In case of any issues, the system defaults to the dark theme, ensuring no disruption to the user experience.

### Additional Notes:
 - **Future Work**:
     - **Light Theme Creation (Issue #2935)**: The next step will involve the creation of a new `light-theme.css` file, which will house the variables and styles specific to the light theme.
     - The light-mode.css file used in this PR is a placeholder.

### CODE REVIEW

**Comment by mathjazz** [comment]
Nice work!

Please note that the spec has now been updated with the theme selector mockups:
https://github.com/mozilla/pontoon/commit/96051544b3badd7cd0b31da3d1d58c1281474081

I'll share the corresponding HTML and CSS with you.

We should also change the tooltips as noted in the spec.

File: pontoon/base/templates/base.html [hunk 1]
```diff
@@ -33,9 +33,9 @@
   </head>
 
   <body
-    class="{% block class %}{% endblock %}"
+    class="{% block class %}{% endblock %} {% if request.user.is_authenticated %}{{ request.user.profile.theme }}-theme{% endif %}"
     {% if csrf_token %}data-csrf="{{ csrf_token }}"{% endif %}
-  >
+>
```
Comment (mathjazz) [inline pontoon/base/templates/base.html:38]: Please revert changes to this line.

File: pontoon/base/templates/django/base.html [hunk 2]
```diff
@@ -18,7 +18,7 @@
     <link rel="stylesheet" href="{% static 'css/fonts.css' %}" title="" type="text/css" />
     <link rel="stylesheet" href="{% static 'css/boilerplate.css' %}" title="" type="text/css" />
     <link rel="stylesheet" href="{% static 'css/style.css' %}" title="" type="text/css" />
-    <link rel="stylesheet" href="{% static 'css/dark-theme.css' %}" title="" type="text/css" />
+
```
Comment (mathjazz) [inline pontoon/base/templates/django/base.html:21]: Please revert changes to this line.

File: pontoon/base/templates/django/base.html [hunk 3]
```diff
@@ -121,5 +121,6 @@
     <script type="text/javascript" charset="utf-8" src="{% static 'js/lib/jquery-3.6.1.js' %}"></script>
     <script type="text/javascript" charset="utf-8" src="{% static 'js/lib/nprogress.js' %}"></script>
     <script type="text/javascript" charset="utf-8" src="{% static 'js/main.js' %}"></script>
+    <script type="text/javascript" charset="utf-8" src="{% static 'js/theme-switcher.js' %}"></script>
```
Comment (mathjazz) [inline pontoon/base/templates/django/base.html:124]: Please revert changes to this line.

File: pontoon/contributors/templates/contributors/settings.html [hunk 4]
```diff
@@ -64,6 +64,18 @@ <h3>Personal information</h3>
         </div>
       </section>
 
+   <section class="appearance" >
+        <h3>Appearance</h3>
+        <div class="field">
+            <p class="help">Choose if Pontoon’s appearance should be dark, light, or follow your system’s settings</a></p>
```
Comment (mathjazz) [inline pontoon/contributors/templates/contributors/settings.html:70]: Please make sure the text matches the spec (we dropped "Pontoon").

File: pontoon/contributors/templates/contributors/settings.html [hunk 5]
```diff
@@ -64,6 +64,18 @@ <h3>Personal information</h3>
...
+        </div>
+        <span class="toggle-button">
+            <button type="button" class="dark active" title="Use a dark theme in Pontoon"><i class="icon far fa-moon"></i>Dark theme</button>
+            <button type="button" class="light" title="Use a light theme in Pontoon"><i class="icon fa fa-sun"></i>Light theme</button>
+            <button type="button" class="system" title="Use a theme in Pontoon that matches your system settings"><i class="icon fa fa-laptop"></i>Match system</button>
```
Comment (mathjazz) [inline pontoon/contributors/templates/contributors/settings.html:75]: Please make sure the text matches the spec. We dropped "Pontoon". Please also do so with the other two themes.

File: pontoon/contributors/templates/contributors/settings.html [hunk 6]
```diff
@@ -160,4 +172,5 @@ <h3><span class="small stress">Preferred locales to get suggestions from</span><
 
 {% block extend_js %}
   {% javascript 'settings' %}
+  {% javascript 'theme-switcher' %}
```
Comment (mathjazz) [inline pontoon/contributors/templates/contributors/settings.html:175]: The theme-switcher.js is already loaded as part of the base template and we also don't have a 'theme-switcher' defined in PIPELINE_JS, so let's revert this line.

File: pontoon/contributors/views.py [hunk 7]
```diff
@@ -187,6 +187,36 @@ def toggle_user_profile_attribute(request, username):
...
+        )
+
+    theme = request.POST.get("theme", None)
+    allowed_themes = ["light", "dark", "system"]
+
+    if theme not in allowed_themes:
+        return JsonResponse(
+            {"status": False, "message": "Bad Request: Invalid theme"},
+            status=400,
+        )
```
Comment (mathjazz) [inline pontoon/contributors/views.py:211]: We should only have themes defined in one places. [Looks like](https://stackoverflow.com/questions/60351804/no-validation-on-field-choices-django-postgres) calling `full_clean()` before `save()` will do the validation "for free"?

File: pontoon/contributors/views.py [hunk 8]
```diff
@@ -250,7 +280,6 @@ def settings(request):
             request.POST,
             instance=profile,
         )
-
```
Comment (mathjazz) [inline pontoon/contributors/views.py:253]: Please revert changes to this line.

**Comment by mathjazz** [comment]
Excellent work with the updates! Added a few more minor requests.

File: pontoon/contributors/urls.py [hunk 9]
```diff
@@ -62,6 +62,12 @@ class UsernameConverter(StringConverter):
         views.toggle_user_profile_attribute,
         name="pontoon.contributors.toggle_user_profile_attribute",
     ),
+    # API: Toggle user theme preference
+    path(
+        "api/v1/user/theme/<username:username>/",
```
Comment (mathjazz) [inline pontoon/contributors/urls.py:67]: We should instead use `"api/v1/user/<username:username>/theme/"` (user identifier should follow the "user").

File: pontoon/contributors/templates/contributors/settings.html [hunk 10]
```diff
@@ -64,6 +64,18 @@ <h3>Personal information</h3>
         </div>
       </section>
 
+   <section class="appearance" >
+        <h3>Appearance</h3>
+        <div class="field">
+            <p class="help">Choose if appearance should be dark, light, or follow your system’s settings</a></p>
+        </div>
+        <span class="toggle-button">
+            <button type="button" class="dark active" title="Use a dark theme"><i class="icon far fa-moon"></i>Dark theme</button>
```
Comment (mathjazz) [inline pontoon/contributors/templates/contributors/settings.html:73]: We need to dynamically set `active` instead of hard-coding it.

Inspiration:
https://github.com/mozilla/pontoon/blob/main/pontoon/base/templates/widgets/toggle.html

**Comment by mathjazz** [comment]
Thanks for another set of updates!

I had a closer look at how the toggle logic actually works. Here's my list of proposals:

1. System theme does not get applied. Themes are applied by setting the corresponding class on the `<body>` element, which results in the use of the right theme file. Since we don't have a `system-theme.css` file, setting `system-theme` on the `<body>` element doesn't apply any of the theme files. If the system theme is set, we should [detect what it is](https://stackoverflow.com/questions/56393880/how-do-i-detect-dark-mode-using-javascript) and set `light-theme` / `dark-theme` class accordingly. It would also be nice to add an event handler (as used in the Stackoverflow) that will change the theme immediately after it is changed in the OS.

1. Changing the theme doesn't seem to apply in the Translate view.

1. In order to prevent the error triggered on theme switch, you'll need to make the selector at https://github.com/mozilla/pontoon/blob/main/pontoon/contributors/static/js/settings.js#L3 more specific, to make sure it doesn't apply to the theme selector.

1. Changes to the `specs/` folder need to be excluded from the PR. You'll need to rebase against the `main` branch.

File: pontoon/base/static/js/themeSwitcher.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
+$(function () {
+  function getSystemTheme() {
+    if (
+      window.matchMedia &&
+      window.matchMedia('(prefers-color-scheme: dark)').matches
+    ) {
+      return 'dark';
+    } else {
+      return 'light';
```
Comment (mathjazz) [inline pontoon/base/static/js/themeSwitcher.js:9]: We should default to dark, not light. So if the user sets "System" in Pontoon and "Auto" on macOS, she should get the dark theme. Right now she gets the light theme.

File: pontoon/base/static/js/themeSwitcher.js [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    }
+  }
+
+  function updateTheme(newTheme) {
```
Comment (mathjazz) [inline pontoon/base/static/js/themeSwitcher.js:13]: Maybe `applyTheme()` is more accurate?

File: pontoon/base/static/js/themeSwitcher.js [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    if (newTheme === 'system') {
+      newTheme = getSystemTheme();
+    }
+    $('body')
+      .removeClass('dark-theme light-theme system-theme')
+      .addClass(`${newTheme}-theme`);
+
+    if (newTheme !== 'system') {
+      Pontoon.endLoader(`Theme changed to ${newTheme}.`);
```
Comment (mathjazz) [inline pontoon/base/static/js/themeSwitcher.js:22]: I would move the notification down to the `ajax.success()` call and call it regardless of the theme being set (i.e. remove the condition).

File: pontoon/base/static/js/themeSwitcher.js [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    }
+  }
+
+  window
+    .matchMedia('(prefers-color-scheme: dark)')
+    .addEventListener('change', function (e) {
+      if ($('.appearance .toggle-button button.system').hasClass('active')) {
```
Comment (mathjazz) [inline pontoon/base/static/js/themeSwitcher.js:29]: This will work on the /settings page. But what if another page is opened when the user changes system settings? We could for example store the actual user theme setting in a data-theme attribute on the body element and use that to check if the system theme is enabled.

File: pontoon/base/static/js/themeSwitcher.js [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+        if (e.matches) {
+          updateTheme('dark');
+        } else {
+          updateTheme('light');
+        }
+      }
+    });
+
+  $(document).ready(function () {
```
Comment (mathjazz) [inline pontoon/base/static/js/themeSwitcher.js:38]: We are already in the `$(document).ready()` as of the first line of this file, so you can remove this wrapper:
https://learn.jquery.com/using-jquery-core/document-ready/

File: pontoon/base/static/js/themeSwitcher.js [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+  $('.appearance .toggle-button button').click(function (e) {
+    e.preventDefault();
+
+    var self = $(this);
+
+    if (self.is('.active')) {
+      return;
+    }
+
+    var theme = self.attr('class').split(' ')[0];
```
Comment (mathjazz) [inline pontoon/base/static/js/themeSwitcher.js:54]: This looks fragile. I suggest we also put values in button `value` attributes and then just use `self.val()` here.

File: pontoon/base/templates/base.html [hunk 17]
```diff
@@ -33,9 +33,10 @@
   </head>
 
   <body
-    class="{% block class %}{% endblock %}"
+    class="{% block class %}{% endblock %} {% if request.user.is_authenticated %}{% if user.profile.theme == 'system' %}system{% else %}{{ user.profile.theme }}{% endif %}-theme{% endif %}"
```
Comment (mathjazz) [inline pontoon/base/templates/base.html:36]: Isn't this the same as this?

```jinja2
<body class="{% block class %}{% endblock %} {% if request.user.is_authenticated %}{{ user.profile.theme }}-theme{% endif %}">
```

File: pontoon/settings/base.py [hunk 18]
```diff
@@ -515,11 +516,12 @@ def _default_from_email():
             "js/lib/jquery.color-2.1.2.js",
             "js/lib/nprogress.js",
             "js/main.js",
+            "js/themeSwitcher.js",
```
Comment (mathjazz) [inline pontoon/settings/base.py:519]: We use kebab-case filenames for static files in `/pontoon`.

File: pontoon/settings/base.py [hunk 19]
```diff
@@ -515,11 +516,12 @@ def _default_from_email():
...
         ),
         "output_filename": "js/base.min.js",
     },
     "translate": {
-        "source_filenames": ("translate.js",),
+        "source_filenames": ("translate.js", "js/themeSwitcher.js"),
```
Comment (mathjazz) [inline pontoon/settings/base.py:524]: We don't need that code in the translate view. We should re-implement the logic in React.

File: pontoon/translate/views.py [hunk 20]
```diff
@@ -64,6 +68,7 @@ def translate(request, locale, project, resource):
     context = {
         "locale": get_preferred_locale(request),
         "notifications": [],
+        "user_theme": user_theme,
```
Comment (mathjazz) [inline pontoon/translate/views.py:71]: This will fail if the user is not authenticated. Do we even use it?

File: translate/public/translate.html [hunk 21]
```diff
@@ -16,15 +16,18 @@
     <title>Pontoon</title>
 
     <!-- Inlining CSS mitigates/prevents fouc. -->
-    <style> body { background: #3f4752 !important; } </style>
+    <style> body.dark-theme { background: #3f4752 !important; } </style>
```
Comment (mathjazz) [inline translate/public/translate.html:19]: What is the purpose of this change?

File: translate/public/translate.html [hunk 22]
```diff
@@ -16,15 +16,18 @@
...
     <link href="/static/css/fontawesome-all.css" rel="stylesheet" />
     <link href="/static/css/boilerplate.css" rel="stylesheet" />
     <link href="/static/css/fonts.css" rel="stylesheet" />
+    <link href="/static/css/dark-theme.css" rel="stylesheet" />
+    <link href="/static/css/light-theme.css" rel="stylesheet" />
```
Comment (mathjazz) [inline translate/public/translate.html:24]: Why not include these files in PIPELINE_CSS? We already include dark-theme.css there.

File: translate/public/translate.html [hunk 23]
```diff
@@ -16,15 +16,18 @@
...
     {% stylesheet 'translate' %}
 
     {% include "tracker.html" %}
   </head>
-  <body>
+  <body class="{% block class %}{% endblock %} {% if request.user.is_authenticated %}{% if user.profile.theme == 'system' %}system{% else %}{{ user.profile.theme }}{% endif %}-theme{% endif %}"
```
Comment (mathjazz) [inline translate/public/translate.html:29]: See the base.html on how we could possible simplify this line.

File: translate/public/translate.html [hunk 24]
```diff
@@ -16,15 +16,18 @@
...
+  {% if csrf_token %}data-csrf="{{ csrf_token }}"{% endif %}>
```
Comment (mathjazz) [inline translate/public/translate.html:30]: No need for the `csrf_token` here.

File: translate/public/translate.html [hunk 25]
```diff
@@ -34,6 +37,6 @@
         data-notifications="{{ notifications | to_json }}"
         translate="no"
     ></div>
-    {% javascript 'translate' %}
+    {% javascript 'translate' %}
```
Comment (mathjazz) [inline translate/public/translate.html:40]: Please revert this change.

File: pontoon/base/static/js/theme-switcher.js [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  function applyTheme(newTheme) {
+    if (newTheme === 'system') {
+      newTheme = getSystemTheme();
+    }
+    $('body')
+      .removeClass('dark-theme light-theme system-theme')
+      .addClass(`${newTheme}-theme`)
+      .data('theme', newTheme);
```
Comment (mathjazz) [inline pontoon/base/static/js/theme-switcher.js:20]: We should move this line to the `ajax.success` function. `applyTheme()` is also called from the event handler, which should not set the data attribute.

Comment (mathjazz) [reply pontoon/base/static/js/theme-switcher.js:20]: And, we should also set the `body[data-theme]` attribute in the templates, otherwise changing the system theme in the OS doesn't work.

File: pontoon/base/static/js/theme-switcher.js [hunk 27]
```diff
@@ -0,0 +1,79 @@
+$(function () {
+  function getSystemTheme() {
+    if (
+      window.matchMedia &&
+      window.matchMedia('(prefers-color-scheme: light)').matches
```
Comment (mathjazz) [inline pontoon/base/static/js/theme-switcher.js:5]: We should change this back to dark (or switch the returns). Sorry for misleading you - the Auto setting on macos changes based on time and has no effect here.

File: pontoon/contributors/templates/contributors/settings.html [hunk 28]
```diff
@@ -64,6 +64,18 @@ <h3>Personal information</h3>
         </div>
       </section>
 
+   <section class="appearance" >
```
Comment (mathjazz) [inline pontoon/contributors/templates/contributors/settings.html:67]: Mind the extra space at the end.

File: pontoon/base/templates/base.html [hunk 29]
```diff
@@ -33,8 +33,9 @@
   </head>
 
   <body
-    class="{% block class %}{% endblock %}"
+    class="{% block class %}{% endblock %} {% if request.user.is_authenticated %}{{ user.profile.theme }}-theme{% endif %}"
     {% if csrf_token %}data-csrf="{{ csrf_token }}"{% endif %}
+    data-theme="{% if user.profile.theme %}{{ user.profile.theme }}{% else %}system{% endif %}">
```
Comment (mathjazz) [inline pontoon/base/templates/base.html:38]: Mind the > at the end of the line.

**Review by mathjazz — COMMENTED** [review]
Just a couple of comments left!

What is the status of applying the theme in the translate view?

Please also look into the failing tests.

**Comment by mathjazz** [comment]
Just noticed that the insights pages (e.g. http://localhost:8000/sl/insights/ and http://localhost:8000/projects/firefox/insights/ appear to be broken (JS error).

**Comment by mathjazz** [comment]
> Just noticed that the insights pages (e.g. http://localhost:8000/sl/insights/ and http://localhost:8000/projects/firefox/insights/ appear to be broken (JS error).

To fix that, we should use `document.body` instead of `document.documentElement` in places like:
https://github.com/mozilla/pontoon/blob/main/pontoon/insights/static/js/insights_tab.js#L7

And also get rid of `trim()` after `style.getPropertyValue()`.

File: pontoon/base/templates/base.html [hunk 30]
```diff
@@ -33,9 +33,9 @@
   </head>
 
   <body
-    class="{% block class %}{% endblock %}"
+    class="{% block class %}{% endblock %} {% if request.user.is_authenticated %}{{ user.profile.theme }}{% else %}dark{% endif %}-theme"
```
Comment (mathjazz) [inline pontoon/base/templates/base.html:36]: We now repeat this very same line 4 times in total:
```
{% if request.user.is_authenticated %}{{ user.profile.theme }}{% else %}dark{% endif %}
```

It would be nice to factor it out, perhaps into a [custom template tag](https://github.com/mozilla/pontoon/blob/05b851058a0f33d5fc60ef528471d9490898bac0/pontoon/base/templatetags/helpers.py#L30-L36), so we'd just call something like` {{ theme(request.user) }}`.

File: translate/src/context/Theme.tsx [hunk 31]
```diff
@@ -16,36 +16,44 @@ function getSystemTheme() {
...
 
 export function ThemeProvider({ children }: { children: React.ReactElement }) {
-  const [theme] = useState(() => {
-    const currentTheme = document.body.getAttribute('data-theme') || 'system';
-    if (currentTheme === 'system') {
-      return getSystemTheme();
-    }
-    return currentTheme;
-  });
+  const [theme] = useState(() => document.body.getAttribute('data-theme')!);
```
Comment (mathjazz) [inline translate/src/context/Theme.tsx:19]: Why is there `!` at the end of the line here and in a couple of places below? There are warning if you run `make eslint`.

File: translate/src/context/Theme.tsx [hunk 32]
```diff
@@ -16,36 +16,44 @@ function getSystemTheme() {
...
-        document.body.getAttribute('data-theme') || 'system';
+      let userThemeSetting = document.body.getAttribute('data-theme')!;
 
       if (userThemeSetting === 'system') {
-        applyTheme(e.matches ? 'dark' : 'light');
+        if (e.matches) {
+          applyTheme('dark');
+        } else {
+          applyTheme('light');
+        }
```
Comment (mathjazz) [inline translate/src/context/Theme.tsx:43]: Actually what I wanted to say is that I like the previous version and we should use it in both files. :-)))

File: translate/src/context/Theme.tsx [hunk 33]
```diff
@@ -16,36 +16,44 @@ function getSystemTheme() {
...
     mediaQuery.addEventListener('change', handleThemeChange);
-
-    applyTheme(theme);
+    if (document.body.classList.contains('system-theme')) {
+      let systemTheme = getSystemTheme();
+      document.body.classList.remove('system-theme');
+      document.body.classList.add(`${systemTheme}-theme`);
+    } else {
+      applyTheme(document.body.getAttribute('data-theme')!);
+    }
```
Comment (mathjazz) [inline translate/src/context/Theme.tsx:56]: Could you explain why did you make this change?

**Comment by mathjazz** [comment]
Nice work with the updates!

* I've kicked the pytest - I only have 3 failing tests locally.
* As discussed, there is a JS error on the /profile when the system theme is selected. I'll take a look.

File: translate/src/context/Theme.tsx (hunk 31 above)
Comment (ayanaar) [reply translate/src/context/Theme.tsx:19]: I did this because of the expectation that the data-theme attribute will always be present on the document.body and thus will not return a null. I used it to bypass the TypeScript type-checking as document.body.getAttribute('data-theme') can potentially return null, and I get an error otherwise. Should I explicitly set a default value instead?

File: translate/src/context/Theme.tsx (hunk 33 above)
Comment (ayanaar) [reply translate/src/context/Theme.tsx]: Removed, this was a previous iteration of the code.

**Review by mathjazz — COMMENTED** [review]
Great work - the core functionality is there!

Pytest is failing due to a conflicting migration. We have recently landed an existing migration 0047, so you'll need to rename yours to 0048 and update the dependency migration: [pontoon/base/static/css/light-theme.css](https://github.com/mozilla/pontoon/tree/main/pontoon/base/migrations).

We also need to get rid of the eslint warnings.

File: translate/src/context/Theme.tsx (hunk 31 above)
Comment (mathjazz) [reply translate/src/context/Theme.tsx:19]: Yeah, seems like we'll need the explicit value anyways.

**Review by mathjazz — COMMENTED** [review]
Excellent job!

Now we need to fix the tests and possibly rebase and we're ready to merge.

**Comment by mathjazz** [comment]
As discussed on Slack, we should probably revert the last patch and remove "theme" from the form.

As for the rebase, I suggest you create a fresh branch out of the current `main`, `git apply` [the patch](https://github.com/mozilla/pontoon/pull/2972.patch), remove the changes in the `specs` folder and push to a new PR.

**Comment by mathjazz** [comment]
Works like a charm! Woohoo! 🥳 

Continuing in #2996.
